src/python/pants/bsp/spec.py

```diff
@@ -3,14 +3,29 @@
 from __future__ import annotations
 
 from dataclasses import dataclass
-from enum import IntEnum
-from typing import Any
+from enum import Enum, IntEnum
+from typing import Any, ClassVar
 
 from pants.bsp.utils import freeze_json
 from pants.build_graph.address import Address, AddressInput
 
 # -----------------------------------------------------------------------------------------------
-# Base types
+# Base classes
+# -----------------------------------------------------------------------------------------------
+
+
+class BSPNotification:
+    """Base class for all notifications so that a notification carries its RPC method name."""
+
+    notification_name: ClassVar[str]
+
+    def to_json_dict(self) -> dict[str, Any]:
+        ...
```

Comment:
Better is to raise `NotImplementedError` I think. `...` is meant for `.pyi` type stub files.

Reply:
Will do.

Reply:
@Eric-Arellano Generally speaking, [`...` is an OK convention for stubbed methods](https://twitter.com/raymondh/status/1430540901772898306), as long as they're going to get written

Reply:
This method is intended as an abstract method to be implemented by subclasses. Would ABCMeta be better?

Reply:
Generally, yeah. But John pointed out he's given up on ABC in most cases because MyPy + dataclasses + ABC is buggy and seems overkill.

Reply:
So `raise NotImplemented` then?

Reply:
I'd say yeah, but don't burn a tree to fix this. Chris has a point it's not a big deal to use `...`. Nbd